Support dynamic quantum architecture for IQM QPUs (#3298)
NVIDIA/cuda-quantum: docs/sphinx/using/backends/hardware/superconducting.rst

.. _iqm-backend:

The `IQM Resonance <https://meetiqm.com/products/iqm-resonance/>`__ portal offers access to various different IQM quantum computers.
The machines available will be constantly extended as development progresses.
Programmers of CUDA-Q may access the IQM Server from either C++ or Python.

With this version it is no longer necessary to define the target QPU architecture in the code or at compile-time.
The IQM backend integration now contacts at runtime the configured IQM server and fetches the active dynamic quantum architecture of the QPU.
This is then used as input to transpile the quantum kernel code just-in-time for the target QPU topology.
By setting the environment variable ``IQM_SERVER_URL`` the target server can be selected just when executing the program.

> As result the python script or the compiled C++ program can be executed on different QPUs without any changes to the code.

Reviewer: nit: 
```suggestion
As result the python script or the compiled C++ program can be executed on different QPUs without recompilation or code changes.
```

> For IQM architectures two-qubit gates can only be performed on adjacent qubits. For more information, we refer to the respective hardware documentation.

Reviewer: These last two sentences here are outdated and should be removed. For application code, the execution on IQM backends should be supported regardless of what gates are used on which qubits.

Author: Thank you for the confirmation. I was unsure about this since these lines were recently rewritten. So I left them as they were. I will update accordingly.

        .. code:: python

            cudaq.sample(kernel, shots_count=10000)

        To see a complete example for using IQM server backends, take a look at our :doc:`Python examples<../../examples/examples>`.

.. tab:: C++

        To target quantum kernel code for execution on an IQM Server,
        pass the ``--target iqm`` flag to the ``nvq++`` compiler.

        .. code:: bash

            nvq++ --target iqm src.cpp

        Once the binary for an IQM QPU is compiled, it can be executed against any IQM Server:

        .. code:: bash

            nvq++ --target iqm src.cpp -o program
            IQM_SERVER_URL="https://demo.qc.iqm.fi/" ./program

        To emulate the IQM machine locally, without submitting to the IQM Server,
        you can also pass the ``--emulate`` flag to ``nvq++``. This will emit any target
        specific compiler diagnostics, before running a noise free emulation.

        .. code:: bash

>             nvq++ --emulate --target iqm src.cpp

Reviewer: I believe emulation would not work fully as expected - it would not do include the layout for a specific device. 
Let me get back on that with some options.
